pym/bob/input.py

```diff
@@ -2015,10 +2015,12 @@ def prepare(self, pathFormatter, inputEnv, sandboxEnabled, inputStates, inputSan
         # filter indirect packages and add to result list
         tmp = indirectPackages
         indirectPackages = []
+        cache = {}
         for p in tmp:
-            p2 = thisDeps.setdefault(p.getPackage().getName(), p)
+            p2 = cache.setdefault(p.getPackage().getName(), p)
             if p2 is p:
-                results.append(p)
+                if not p in results:
```

Comment:
This (implicitly) uses the variant-id of a step to see if if it is already in the result list. But identical packages with different names are (and were) allowed. Oh, and this is O(n^2) in the worst case.